cachelib/navy/common/FdpNvme.h

```diff
@@ -0,0 +1,271 @@
+// So NVMe data is needed for initialization and IO cmd formation.
+class NvmeData {
+ public:
+  NvmeData() = default;
+  NvmeData& operator=(const NvmeData&) = default;
+
+  explicit NvmeData(int nsId, uint32_t lbaShift, uint64_t nLba,
+      uint32_t maxTfrSize)
+    : nsId_(nsId), lbaShift_(lbaShift), nLba_(nLba), maxTfrSize_(maxTfrSize) {}
+
+  // NVMe Namespace ID
+  int nsId() const { return nsId_;}
+
+  // LBA shift number to calculate blocksize
+  uint32_t lbaShift() const { return lbaShift_;}
+
+  // Num of LBAs in device
+  uint64_t nLba() const { return nLba_;}
+
+  // Get the max transfer size of NVMe device.
+  uint32_t getMaxTfrSize() { return maxTfrSize_; }
+
+ private:
+  int nsId_;
+  uint32_t lbaShift_;
+  uint64_t nLba_;
+  uint32_t maxTfrSize_{};
+};
+
+// FDP specific info and handling
+//
+// This embeds the FDP semantics and specific io-handling.
+// Note: IO with FDP semantics need to be sent through Io_Uring_cmd interface
+// as of now; and not supported through conventional block interfaces.
+class FdpNvme {
+ public:
+  FdpNvme(int fd);
+
+  FdpNvme(const FdpNvme&) = delete;
+  FdpNvme& operator=(const FdpNvme&) = delete;
+
+  // Allocates an FDP specific placement handle. This handle will be
+  // interpreted by the device for data placement.
+  int allocateFdpHandle();
+
+  // Get the max transfer size of NVMe device.
+  uint32_t getMaxTfrSize() { return nvmeData_.getMaxTfrSize(); }
+
+  // Get the NVMe specific info on this device.
+  NvmeData& getNvmeData() { return nvmeData_; }
+
+  // Reads FDP status descriptor into Buffer
+  Buffer nvmeFdpStatus(int fd);
+
+  // Prepares the Uring_Cmd sqe for read command.
+  void prepReadUringCmdSqe(struct io_uring_sqe& sqe, int fd, void* buf,
+      size_t size, off_t start);
+
+  // Prepares the Uring_Cmd sqe for write command with FDP handle.
+  void prepWriteUringCmdSqe(struct io_uring_sqe& sqe, int fd, void* buf,
+      size_t size, off_t start, int handle);
+
+ private:
+  // Prepares the Uring_Cmd sqe for read/write command with FDP directives.
+  void prepFdpUringCmdSqe(struct io_uring_sqe& sqe, int fd, void* buf,
+      size_t size, off_t start, uint8_t opcode, uint8_t dtype,
+      uint16_t dspec, NvmeData& nvmeData);
+
+  // Get FDP PlacementID for a NVMe NS specific PHNDL
+  uint16_t getFdpPID(uint16_t fdpPHNDL) {
+    // Fetch PID like placementIDs_[placementHandle].
+    return placementIDs_[fdpPHNDL];
+  }
+
+  // Reads NvmeData for a NVMe device
+  NvmeData readNvmeInfo(int fd);
+
+  // Initialize the FDP device and populate necessary info.
+  void initializeFDP(int fd);
+
+  // Generic NVMe IO mgmnt receive cmd
+  int nvmeIOMgmtRecv(int fd, uint32_t nsid, void *data, uint32_t data_len,
+      uint16_t mos, uint8_t mo);
+
+  // 0u is considered as the default placement ID
+  static constexpr uint16_t kDefaultPIDIdx = 0u;
+
+  // The mapping table of <PHNDL, PID> of Namespaces
+  std::map<uint16_t, uint16_t> placementIDs_{};
```

Comment:
Why map? Just `std::vector<uint16_t>` would work.

Reply:
yes, if we are returning phndl instead of pid. Thanks for pointing out that.